src/search_unused.rs

```diff
@@ -343,72 +343,72 @@ pub(crate) fn find_unused(
-                            // Otherwise, take everything up to the #.
-                            temp = subset.split('#').next();
-                        }
-                    }
-
-                    temp
-                };
-
-                package_found.map_or(false, |node_package_name| node_package_name == package_name)
-            })
-            .expect("the current package must be in the dependency graph")
-            .deps
-            .iter()
-            .map(|node_dep| node_dep.name.clone())
-            .collect()
+        let resolve = metadata.resolve.as_ref().unwrap();
+        if let Some(ref root) = resolve.root {
+            // This gives us resolved dependencies, in crate form
+            let root_node = resolve
+                .nodes
+                .iter()
+                .find(|node| node.id == *root)
+                .expect("root should be resolved by cargo-metadata");
+            // This gives us the original dependency table
+            // May have more than resolved if some were never enabled
+            let root_package = metadata
+                .packages
+                .iter()
+                .find(|pkg| pkg.id == *root)
+                .expect("root should appear under cargo-metadata packages");
+            // For every resolved dependency:
+            // look it up in the package list to find the name (the one in registries)
+            // look up that name in dependencies of the root_package;
+            // find if it uses a different key through the rename field
+            root_node
+                .deps
+                .iter()
+                .map(|dep| {
+                    let crate_name = dep.name.clone();
+                    let dep_pkg = metadata
+                        .packages
+                        .iter()
+                        .find(|pkg| pkg.id == dep.pkg)
+                        .expect(
+                            "resolved dependencies should appear under cargo-metadata packages",
+                        );
+                    let mut dep_spec_it = root_package.dependencies.iter().filter(|dep_spec| {
+                        dep_spec.name == dep_pkg.name
+                            && dep_spec.source.as_ref() == dep_pkg.source.as_ref().map(|s| &s.repr)
+                    });
+                    // The dependency can appear more than once, for example if it is both
+                    // a dependency and a dev-dependency (often with more features enabled).
+                    // We'll assume cargo enforces consistency.
+                    let dep_spec = dep_spec_it
+                        .next()
+                        .expect("resolved dependency should have a matching dependency spec");
+                    // If the dependency was renamed, through key = { package = … },
+                    // the original key is in dep_spec.rename
+                    let dep_key = dep_spec.rename.clone().unwrap_or(dep_spec.name.clone());
```

Comment:
nit: `unwrap_or_else(|| dep_spec.name.clone())` to avoid the eager cloning of `dep_spec.name`